erpnext/selling/doctype/sales_order/sales_order.js

```diff
@@ -226,10 +226,43 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
 						fields: fields,
 						primary_action: function() {
 							var data = d.get_values();
+
+							data_object_items = [];
+							for(i = 0; i < data["items"].length; i++) {
+								data_object_items.push(data.items[i].sales_order_item);
+							}
+
+							visible_rows = $(".modal-dialog:has(div[data-fieldtype=\"Table\"][data-fieldname=\"items\"]) .grid-body .grid-row");
+
+							visible_items = [];
+
+							$(visible_rows).each(function() {
+								visible_items.push($(this).find("div[data-fieldname=\"sales_order_item\"] .static-area.ellipsis").text());
+							});
+
+							var selected_data = {};
+							var item_data = [];
+							for (i = 0; i < data["items"].length; i++) {
+								for (j = 0; j < visible_items.length; j++) {
+									if (data.items[i].sales_order_item == visible_items[j]) {
+										item_data.push({
+											"bom": data.items[i].bom,
+											"idx": data.items[i].idx,
+											"item_code": data.items[i].item_code,
+											"pending_qty": data.items[i].pending_qty,
+											"sales_order_item": data.items[i].sales_order_item,
+											"warehouse": data.items[i].warehouse
+										});
+									}
+								}
+							}
```

Comment:
Just a small nitpick but arguably you could cut down in complexity if you just filtered your data["items"] instead of having the two ugly for loops:

```
data["items"]
    .filter(item => visible_items.find(item.sales_order_item))
    .forEach((item, i) => {
        ... rest of your work here ...
    });
```

The intent is much more clear there(I think)